jablib/src/main/java/org/jabref/logic/integrity/RefChecker.java

```diff
@@ -0,0 +1,258 @@
+
+        public ReferenceValidity or(ReferenceValidity other) {
+            if (this instanceof Real || other instanceof Fake) {
+                return this;
+            }
+            if (other instanceof Unsure otherUnsure && this instanceof Unsure thisUnsure) {
+                Unsure merge = new Unsure();
+                merge.addAll(thisUnsure);
+                merge.addAll(otherUnsure);
+                return merge;
+            }
+            return other;
+        }
+
+        private ReferenceValidity lazyOr(ReferenceValiditySupplier other) throws FetcherException {
+            if (this instanceof Real) {
+                return this;
+            } else {
+                return or(other.get());
+            }
+        }
+    }
+
+    public static final class Real extends ReferenceValidity {
+        BibEntry matchingReference;
+
+        public Real(BibEntry matchingReference) {
+            this.matchingReference = matchingReference;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            Real real = (Real) o;
+            return Objects.equals(matchingReference, real.matchingReference);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(matchingReference);
+        }
+
+        public BibEntry getMatchingReference() {
+            return matchingReference;
+        }
+    }
+
+    public static final class Unsure extends ReferenceValidity {
+        Set<BibEntry> matchingReferences;
+
+        public Unsure(BibEntry matchingReference) {
+            this.matchingReferences = new HashSet<>(Set.of(matchingReference));
+        }
+
+        private Unsure() {
+            this.matchingReferences = new HashSet<>();
+        }
+
+        void addAll(Unsure other) {
+            this.matchingReferences.addAll(other.matchingReferences);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            Unsure unsure = (Unsure) o;
+            return Objects.equals(matchingReferences, unsure.matchingReferences);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(matchingReferences);
+        }
+
+        public Set<BibEntry> getMatchingReferences() {
+            return matchingReferences;
+        }
+    }
+
+    public static final class Fake extends ReferenceValidity {
+        public boolean equals(Object o) {
+            return o.getClass() == Fake.class;
+        }
+
+        public int hashCode() {
+            return Objects.hashCode(Fake.class);
+        }
```

Comment:
No neeed for an implementation, JDK's default works well in most cases

Override is only necessary if equals is overridden: https://thefinestartist.com/effective-java/09